ios/browser/api/brave_shields/adblock_engine.mm

```diff
@@ -0,0 +1,134 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/ios/browser/api/brave_shields/adblock_engine.h"
+
+#include "base/strings/sys_string_conversions.h"
+#include "brave/base/mac/conversions.h"
+#include "brave/components/adblock_rust_ffi/src/wrapper.h"
+#include "brave/components/brave_shields/common/adblock_domain_resolver.h"
+
+#if !defined(__has_feature) || !__has_feature(objc_arc)
+#error "This file requires ARC support."
+#endif
+
+@interface AdblockEngineMatchResult ()
+@property(nonatomic, readwrite) bool didMatchRule;
+@property(nonatomic, readwrite) bool didMatchException;
+@property(nonatomic, readwrite) bool didMatchImportant;
+@property(nonatomic, readwrite, copy) NSString* redirect;
+@end
+
+@implementation AdblockEngineMatchResult
+- (instancetype)init {
+  if ((self = [super init])) {
+    self.redirect = @"";
+  }
+  return self;
+}
+@end
+
+@implementation AdblockEngine {
+  std::unique_ptr<adblock::Engine> adblock_engine;
+}
+
+- (instancetype)init {
+  return [self initWithRules:@""];
+}
+
+- (instancetype)initWithRules:(NSString*)rules {
+  if ((self = [super init])) {
+    adblock_engine =
+        std::make_unique<adblock::Engine>(base::SysNSStringToUTF8(rules));
+  }
+  return self;
+}
+
+- (AdblockEngineMatchResult*)matchesURL:(NSString*)url
+                                   host:(NSString*)host
+                                tabHost:(NSString*)tabHost
+                           isThirdParty:(bool)isThirdParty
+                           resourceType:(NSString*)resourceType {
+  bool did_match_rule = false;
+  bool did_match_exception = false;
+  bool did_match_important = false;
+  std::string redirect = "";
+  adblock_engine->matches(
+      base::SysNSStringToUTF8(url), base::SysNSStringToUTF8(host),
+      base::SysNSStringToUTF8(tabHost), isThirdParty,
+      base::SysNSStringToUTF8(resourceType), &did_match_rule,
+      &did_match_exception, &did_match_important, &redirect);
+  auto result = [[AdblockEngineMatchResult alloc] init];
```

Comment:
@antonok-edm can make this a mojo type so we can pass it through without copying all the fields?

Reply:
This is mostly just to make the API nicer on Swift side, which doesn't use `inout` arguments as often